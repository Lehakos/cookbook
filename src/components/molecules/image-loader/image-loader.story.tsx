import React from 'react';
import { storiesOf } from '@storybook/react';

import { ImageLoader } from './image-loader';

const stories = storiesOf('Molecules|ImageLoader', module);

stories
  .add('Fixed size', () => (
    <ImageLoader
      width={320}
      height={240}
      alt="nature"
      src="https://placeimg.com/320/240/nature"
    />
  ))
  .add('Responsive', () => (
    <ImageLoader
      responsive
      alt="animals"
      src="https://placeimg.com/1280/960/animals"
    />
  ))
  .add('Get image size', () => (
    <ImageLoader alt="arch" src="https://placeimg.com/640/480/arch" />
  ))
  .add('Fail to load image', () => (
    <ImageLoader alt="arch" src="https://placeimg.com" />
  ));
