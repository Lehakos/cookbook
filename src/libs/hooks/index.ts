export { useControlledProp } from './use-controlled-prop';
export { useElementSize } from './use-element-size';
