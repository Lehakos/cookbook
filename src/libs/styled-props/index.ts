export { typography, TypographyProps } from './typography';
export { layout, LayoutProps } from './layout';
export { flexbox, FlexboxProps } from './flexbox';
