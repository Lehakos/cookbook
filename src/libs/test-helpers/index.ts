export { render } from './render';
export { mockElementWithBoundingRect } from './mock-element-with-bounding-rect';
