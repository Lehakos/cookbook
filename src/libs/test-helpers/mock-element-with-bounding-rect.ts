export const mockElementWithBoundingRect = ({
  width = 100,
  height = 100,
  top = 0,
  left = 0,
}) => {
  const right = left + width;
  const bottom = top + height;

  Object.defineProperties(Element.prototype, {
    offsetLeft: {
      get() {
        return left;
      },
      configurable: true,
    },
    offsetTop: {
      get() {
        return top;
      },
      configurable: true,
    },
    offsetHeight: {
      get() {
        return height;
      },
      configurable: true,
    },
    offsetWidth: {
      get() {
        return width;
      },
      configurable: true,
    },
  });

  Element.prototype.getBoundingClientRect = jest.fn(() => {
    return {
      width,
      height,
      top,
      left,
      bottom,
      right,
    };
  });

  const element = document.createElement('div');

  return element;
};
