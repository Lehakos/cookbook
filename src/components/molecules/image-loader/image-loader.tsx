import React, { useReducer, useEffect } from 'react';

import { ImagePlaceholder, ImagePlaceholderProps } from 'components/atoms';
import { StrictUnion } from 'shared/types';

import * as s from './image-loader.styles';

type Size = {
  width: number | string;
  height: number | string;
};

export type Props = {
  src: string;
  alt: string;
} & ImagePlaceholderProps;

interface State {
  error: boolean;
  loaded: boolean;
  computedSize: StrictUnion<Size | {}>;
}

type LoadImageStartAction = { type: 'loadImageStart' };
type LoadImageSuccessAction = { type: 'loadImageSuccess'; payload: Size };
type LoadImageFailAction = { type: 'loadImageFail' };

const reducer = (
  state: State,
  action: LoadImageStartAction | LoadImageSuccessAction | LoadImageFailAction,
) => {
  switch (action.type) {
    case 'loadImageStart':
      return {
        ...state,
        loaded: false,
        error: false,
      };

    case 'loadImageSuccess':
      return {
        ...state,
        loaded: true,
        computedSize: action.payload,
      };

    case 'loadImageFail':
      return {
        ...state,
        error: true,
      };

    default:
      return state;
  }
};

export const ImageLoader: React.FunctionComponent<Props> = function({
  src,
  alt,
  width,
  height,
  responsive,
}) {
  const initialState: State = {
    loaded: false,
    error: false,
    computedSize:
      typeof width !== 'undefined' && typeof height !== 'undefined'
        ? { width, height }
        : {},
  };
  const [state, dispatch] = useReducer(reducer, initialState);

  const loadImage = () => {
    const currentSrc = src;
    dispatch({ type: 'loadImageStart' });

    const img = new Image();

    const loadHandler = () => {
      if (currentSrc !== src) {
        return;
      }

      const computedSize = {
        width: width || img.naturalWidth,
        height: height || img.naturalHeight,
      };

      dispatch({ type: 'loadImageSuccess', payload: computedSize });
    };

    const errorHandler = (e: ErrorEvent) => {
      if (currentSrc !== src) {
        return;
      }
      if (process.env.NODE_ENV !== 'test') {
        console.error(e.error); // eslint-disable-line
      }
      dispatch({ type: 'loadImageFail' });
    };

    img.addEventListener('load', loadHandler);
    img.addEventListener('error', errorHandler);

    img.src = src;

    return () => {
      img.removeEventListener('load', loadHandler);
      img.removeEventListener('error', errorHandler);
    };
  };

  useEffect(() => {
    return loadImage();
  }, [src]);

  const { loaded, error, computedSize } = state;
  const showImage = !error && loaded;

  return (
    <ImagePlaceholder {...computedSize} responsive={responsive}>
      {error && (
        <div>
          Fail to load image. Here is a text description:
          <p>{alt}</p>
        </div>
      )}
      {showImage && <s.Img src={src} alt={alt} />}
    </ImagePlaceholder>
  );
};
