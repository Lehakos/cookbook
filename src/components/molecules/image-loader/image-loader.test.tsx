import React from 'react';
import { waitForElement, fireEvent } from '@testing-library/react';

import { render } from 'libs/test-helpers';

import { ImageLoader, Props } from './image-loader';

const DESCRIPTION = 'Alternative text';
const LOAD_FAILURE_SRC = 'LOAD_FAILURE_SRC';
const LOAD_SUCCESS_SRC = 'LOAD_SUCCESS_SRC';
const WIDTH = 640;
const HEIGHT = 480;
const NATURAL_WIDTH = 100;
const NATURAL_HEIGHT = 100;

const renderComp = (props: Partial<Props> = {}) => {
  const defaultProps: Props = {
    width: WIDTH,
    height: HEIGHT,
    alt: DESCRIPTION,
    src: LOAD_SUCCESS_SRC,
  };

  return render(<ImageLoader {...defaultProps} {...props} />);
};

describe('<ImageLoader />', () => {
  beforeAll(() => {
    Object.defineProperties((global as any).Image.prototype, {
      src: {
        set(src) {
          if (src === LOAD_FAILURE_SRC) {
            setTimeout(() =>
              fireEvent(
                this,
                new ErrorEvent('error', { error: new Error('some error') }),
              ),
            );
          } else if (src === LOAD_SUCCESS_SRC) {
            setTimeout(() => fireEvent(this, new Event('load')));
          }
        },
        configurable: true,
      },
      naturalWidth: {
        get() {
          return NATURAL_WIDTH;
        },
        configurable: true,
      },
      naturalHeight: {
        get() {
          return NATURAL_HEIGHT;
        },
        configurable: true,
      },
    });
  });

  it('renders properly', () => {
    const { container } = renderComp();
    expect(container.firstChild).toMatchSnapshot();
  });

  describe('when image loaded successfully', () => {
    it('shows the image', async () => {
      const { queryByAltText } = renderComp();
      const el = await waitForElement(() => queryByAltText(DESCRIPTION));

      expect(el).toBeTruthy();
    });
  });

  describe('when image loading failed', () => {
    it('shows alternative text', async () => {
      const { queryByText } = renderComp({
        src: LOAD_FAILURE_SRC,
      });
      const el = await waitForElement(() => queryByText(DESCRIPTION));

      expect(el).toBeTruthy();
    });
  });

  describe('when image is loading', () => {
    it('shows placeholder', () => {
      const { queryByText, queryByAltText } = renderComp();
      const altEl = queryByText(DESCRIPTION);
      const imgEl = queryByAltText(DESCRIPTION);

      expect(altEl).toBe(null);
      expect(imgEl).toBe(null);
    });
  });
});
