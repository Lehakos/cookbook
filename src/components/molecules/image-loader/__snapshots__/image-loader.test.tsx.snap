// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ImageLoader /> renders properly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  width: 640px;
  height: 480px;
  text-align: center;
  background-color: #e2e2e2;
}

.c1 {
  text-transform: uppercase;
  -webkit-letter-spacing: 0.1em;
  -moz-letter-spacing: 0.1em;
  -ms-letter-spacing: 0.1em;
  letter-spacing: 0.1em;
}

<div
  class="c0"
  height="480px"
  width="640px"
>
  <span
    aria-hidden="true"
    class="c1"
  >
    640
     x 
    480
  </span>
</div>
`;
